ufl/algorithms/comparison_checker.py

```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+"""Algorithm to check for 'comparison' nodes
+in a form when the user is in 'complex mode'"""
+
+from ufl.corealg.multifunction import MultiFunction
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algebra import Real
+from ufl.constantvalue import IntValue, FloatValue, Zero
+from ufl.argument import Argument
+
+
+class CheckComparisons(MultiFunction):
+    """Raises an error if comparisons are done with complex quantities.
+
+    If quantities are real, adds the Real operator to the compared quantities.
+
+    Terminals that are real are IntValue, FloatValue, Zero, and Argument
+    (even in complex FEM, the basis functions are real)
+    Operations that produce reals are Abs, Real, Imag.
+    Terminals default to complex, and Sqrt, Pow (defensively) imply complex.
+    Otherwise, operators preserve the type of their operands.
+    """
+    def __init__(self):
+        MultiFunction.__init__(self)
+        self.nodetype = {}
+
+    def expr(self, o, *ops):
+        """Defaults expressions to complex unless they only
+        act on real quantities. Overridden for specific operators.
+
+        Rebuilds objects if necessary.
+        """
+
+        types = {self.nodetype[op] for op in ops}
+
+        if types:
+            t = "complex" if "complex" in types else "real"
+        else:
+            t = "complex"
+
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = t
+        return o
+
+    def gt(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with gt.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def lt(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with lt.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def ge(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with ge.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def le(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with le.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
```

Comment:
Code duplication: same body repeated 4 times.

Reply:
I think factoring it is hardly worth it. It's very little code and factoring is actually less readable.

Reply:
```python
    def compare(self, o, *ops):
        types = {self.nodetype[op] for op in ops}

        if "complex" in types:
            raise ComplexComparisonError("You cannot compare complex numbers.")
        else:
            o = o._ufl_expr_reconstruct_(*map(Real, ops))
            self.nodetype[o] = "bool"
            return o

    gt = compare
    lt = compare
    ge = compare
    le = compare
```

Reply:
Fine. I don't actually think this is an improvement (and the error message is actually incorrect). However anything for  a quiet life.

Reply:
You could write something like "no ordering is defined on complex numbers" for example.